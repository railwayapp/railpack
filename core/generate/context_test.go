package generate

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/railwayapp/railpack/core/app"
	"github.com/railwayapp/railpack/core/config"
	"github.com/railwayapp/railpack/core/logger"
	"github.com/railwayapp/railpack/core/plan"
	"github.com/stretchr/testify/require"
)

type TestProvider struct{}

func (p *TestProvider) Plan(ctx *GenerateContext) error {
	// mise
	mise := ctx.GetMiseStepBuilder()
	nodeRef := mise.Default("node", "18")
	mise.Version(nodeRef, "18", "test")

	// commands
	installStep := ctx.NewCommandStep("install")
	installStep.AddCommand(plan.NewExecCommand("npm install", plan.ExecOptions{}))
	installStep.AddInput(mise.GetLayer())
	installStep.Secrets = []string{}

	buildStep := ctx.NewCommandStep("build")
	buildStep.AddCommand(plan.NewExecCommand("npm run build", plan.ExecOptions{}))
	buildStep.AddInput(plan.NewStepLayer(installStep.Name()))

	ctx.Deploy.DeployInputs = []plan.Layer{
		plan.NewStepLayer(buildStep.Name()),
	}

	return nil
}

func CreateTestContext(t *testing.T, path string) *GenerateContext {
	t.Helper()

	userApp, err := app.NewApp(path)
	require.NoError(t, err)

	env := app.NewEnvironment(nil)
	config := config.EmptyConfig()

	ctx, err := NewGenerateContext(userApp, env, config, logger.NewLogger())
	require.NoError(t, err)

	return ctx
}

func TestGenerateContext(t *testing.T) {
	ctx := CreateTestContext(t, "../../examples/node-npm")
	provider := &TestProvider{}
	require.NoError(t, provider.Plan(ctx))

	// User defined config
	configJSON := `{
		"packages": {
			"node": "20.18.2",
			"go": "1.23.5",
			"python": "3.13.1"
		},
		"aptPackages": ["curl"],
		"steps": {
			"build": {
				"commands": ["echo building"]
			}
		},
		"secrets": ["RAILWAY_SECRET_1", "RAILWAY_SECRET_2"],
		"deploy": {
			"startCommand": "echo hello",
			"variables": {
				"HELLO": "world"
			}
		}
	}`

	var config config.Config
	require.NoError(t, json.Unmarshal([]byte(configJSON), &config))

	ctx.Config = &config

	buildPlan, _, err := ctx.Generate()
	require.NoError(t, err)

	buildPlanJSON, err := json.MarshalIndent(buildPlan, "", "  ")
	require.NoError(t, err)

	var actualPlan map[string]interface{}
	require.NoError(t, json.Unmarshal(buildPlanJSON, &actualPlan))

	serializedPlan, err := json.MarshalIndent(actualPlan, "", "  ")
	require.NoError(t, err)

	snaps.MatchJSON(t, serializedPlan)
}

func TestGenerateContextDockerignore(t *testing.T) {
	t.Run("context with dockerignore", func(t *testing.T) {
		ctx := CreateTestContext(t, "../../examples/dockerignore")

		// Verify dockerignore was parsed during context creation
		require.NotNil(t, ctx.dockerignoreCtx)

		// Test NewLocalLayer with dockerignore patterns
		layer := ctx.NewLocalLayer()
		require.True(t, layer.Local)
		require.NotNil(t, layer.Filter)

		// Should have exclude patterns from .dockerignore
		require.NotEmpty(t, layer.Exclude)
		require.Contains(t, layer.Exclude, ".vscode")
		require.Contains(t, layer.Exclude, "*.log")
		require.Contains(t, layer.Exclude, "__pycache__") // Trailing slash is stripped by parser

		// Should have default include pattern
		require.Equal(t, []string{"."}, layer.Include)
	})

	t.Run("context without dockerignore", func(t *testing.T) {
		ctx := CreateTestContext(t, "../../examples/node-npm")

		// Verify dockerignore context exists but has no patterns
		require.NotNil(t, ctx.dockerignoreCtx)

		// Test NewLocalLayer without dockerignore patterns
		layer := ctx.NewLocalLayer()
		require.True(t, layer.Local)

		// Should use default behavior when no dockerignore patterns exist
		require.NotNil(t, layer.Filter)
		require.Equal(t, []string{"."}, layer.Include)
		require.Empty(t, layer.Exclude)
	})

	t.Run("context creation with no dockerignore", func(t *testing.T) {
		// Test with a directory that exists but has no .dockerignore file
		ctx := CreateTestContext(t, "../../examples/node-npm")

		// Should succeed even without .dockerignore file
		require.NotNil(t, ctx)
		require.NotNil(t, ctx.dockerignoreCtx)

		// Verify parsing works with no file present
		excludes, includes, _ := ctx.dockerignoreCtx.Parse()
		require.Nil(t, excludes)
		require.Nil(t, includes)
	})

	t.Run("context creation fails with invalid dockerignore", func(t *testing.T) {
		// Create a temporary directory with an inaccessible .dockerignore
		tempDir, err := os.MkdirTemp("", "dockerignore-test")
		require.NoError(t, err)
		defer func() {
			if err := os.RemoveAll(tempDir); err != nil {
				t.Errorf("failed to remove temp dir: %v", err)
			}
		}()

		dockerignorePath := filepath.Join(tempDir, ".dockerignore")
		err = os.WriteFile(dockerignorePath, []byte("*.log\nnode_modules\n"), 0644)
		require.NoError(t, err)

		// Make the file unreadable to simulate a parsing error
		err = os.Chmod(dockerignorePath, 0000)
		require.NoError(t, err)
		defer func() { _ = os.Chmod(dockerignorePath, 0644) }()

		// Create app with the temp directory
		userApp, err := app.NewApp(tempDir)
		require.NoError(t, err)

		env := app.NewEnvironment(nil)
		config := config.EmptyConfig()

		// Context creation should fail due to dockerignore parsing error
		ctx, err := NewGenerateContext(userApp, env, config, logger.NewLogger())
		require.Error(t, err)
		require.Contains(t, err.Error(), "failed to parse .dockerignore")
		require.Nil(t, ctx)
	})
}
