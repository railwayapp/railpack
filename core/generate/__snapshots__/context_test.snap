
[TestGenerateContext - 1]
{
 "deploy": {
  "base": {
   "image": "ghcr.io/railwayapp/railpack-runtime:latest"
  },
  "inputs": [
   {
    "step": "build"
   },
   {
    "include": [
     "."
    ],
    "step": "build"
   }
  ],
  "startCommand": "echo hello",
  "variables": {
   "HELLO": "world"
  }
 },
 "secrets": [
  "RAILWAY_SECRET_1",
  "RAILWAY_SECRET_2"
 ],
 "steps": [
  {
   "assets": {
    "mise.toml": "[tools]\n  [tools.go]\n    version = \"1.23.5\"\n  [tools.node]\n    version = \"20.18.2\"\n  [tools.python]\n    version = \"3.13.1\"\n"
   },
   "commands": [
    {
     "path": "/mise/shims"
    },
    {
     "customName": "create mise config",
     "name": "mise.toml",
     "path": "/etc/mise/config.toml"
    },
    {
     "cmd": "sh -c 'mise trust -a \u0026\u0026 mise install'",
     "customName": "install mise packages: go, node, python"
    }
   ],
   "inputs": [
    {
     "image": "ghcr.io/railwayapp/railpack-builder:latest"
    }
   ],
   "name": "packages:mise",
   "variables": {
    "MISE_CACHE_DIR": "/mise/cache",
    "MISE_CONFIG_DIR": "/mise",
    "MISE_DATA_DIR": "/mise",
    "MISE_INSTALLS_DIR": "/mise/installs",
    "MISE_SHIMS_DIR": "/mise/shims"
   }
  },
  {
   "commands": [
    {
     "cmd": "npm install"
    }
   ],
   "inputs": [
    {
     "include": [
      "/mise/shims",
      "/mise/installs",
      "/usr/local/bin/mise",
      "/etc/mise/config.toml",
      "/root/.local/state/mise"
     ],
     "step": "packages:mise"
    }
   ],
   "name": "install"
  },
  {
   "commands": [
    {
     "cmd": "sh -c 'echo building'",
     "customName": "echo building"
    }
   ],
   "inputs": [
    {
     "step": "install"
    }
   ],
   "name": "build",
   "secrets": [
    "*"
   ]
  }
 ]
}
---
